File: src/api/storePnl.ts
import { generatedGet } from './generated/client';

export interface PnlMonth {
  month: string;
  has_actual: boolean;
  has_estimated: boolean;
}

export interface PnlMetrics {
  revenue: number;
  cogs: number;
  gross_margin: number;
  operating_costs: number;
  ebitda: number;
  depreciation: number;
  ebit: number;
}

export interface PnlMonthlyPoint extends PnlMetrics {
  month: string;
  is_estimated: boolean;
}

export interface PnlAnnualPoint extends PnlMetrics {
  year: string;
  store_count: number;
  month_count: number;
  is_estimated: boolean;
}

export interface PnlStoreOption {
  company_name: string;
  site_code: string;
  location: string;
  regional: string;
  scope_company: string | null;
}

export interface PnlStore extends PnlMetrics {
  company: string;
  site_code: string;
  source_site_code: string;
  location: string;
  regional: string;
  has_estimates: boolean;
}

export interface PnlReconciliation {
  month: string;
  pnl_revenue: number;
  retail_sales_gross: number;
  retail_sales_net: number;
  difference_to_net: number;
  pnl_to_net_sales_pct: number | null;
}

export interface PnlOverview {
  start_month: string;
  end_month: string;
  company: string | null;
  site_code: string | null;
  site_company: string | null;
  regional: string | null;
  summary: PnlMetrics;
  monthly: PnlMonthlyPoint[];
  categories: Record<string, number>;
  stores: PnlStore[];
  reconciliation: PnlReconciliation[];
}

export interface PnlPermissions {
  can_view: boolean;
}

export async function getPnlPermissions(): Promise<PnlPermissions> {
  const data = await generatedGet('pnl_permissions_api_store_pnl_permissions_get') as unknown as PnlPermissions;
  if (
    !data
    || typeof data !== "object"
    || typeof data.can_view !== "boolean"
    || Object.keys(data).some((key) => key !== "can_view")
  ) {
    throw new Error("Invalid P&L permissions response");
  }
  return data;
}

export async function getPnlMonths(signal?: AbortSignal): Promise<PnlMonth[]> {
  const data = await generatedGet('months_api_store_pnl_months_get', { signal }) as { months: PnlMonth[] };
  return data.months;
}

export async function getPnlStores(company: string, regional = "", signal?: AbortSignal): Promise<PnlStoreOption[]> {
  const data = await generatedGet('stores_api_store_pnl_stores_get', {
    params: { company: company || undefined, regional: regional || undefined },
    signal,
  }) as { stores: PnlStoreOption[] };
  return data.stores;
}

export async function getPnlRegions(company: string, signal?: AbortSignal): Promise<string[]> {
  const data = await generatedGet('regions_api_store_pnl_regions_get', {
    params: { company: company || undefined },
    signal,
  }) as { regions: string[] };
  return data.regions;
}

export async function getPnlAnnual(
  company: string,
  siteCode: string,
  siteCompany = "",
  regional = "",
  signal?: AbortSignal,
): Promise<PnlAnnualPoint[]> {
  const data = await generatedGet('annual_api_store_pnl_annual_get', {
    params: {
      company: company || undefined,
      site_code: siteCode || undefined,
      site_company: siteCompany || undefined,
      regional: regional || undefined,
    },
    signal,
  }) as { annual: PnlAnnualPoint[] };
  return data.annual;
}

export async function getPnlOverview(
  startMonth: string,
  endMonth: string,
  company: string,
  siteCode = "",
  siteCompany = "",
  regional = "",
  signal?: AbortSignal,
): Promise<PnlOverview> {
  const data = await generatedGet('overview_api_store_pnl_overview_get', {
    params: {
      start_month: startMonth,
      end_month: endMonth,
      company: company || undefined,
      site_code: siteCode || undefined,
      site_company: siteCompany || undefined,
      regional: regional || undefined,
    },
    signal,
  }) as unknown as PnlOverview;
  return data;
}

File: backend/routers/store_pnl.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from pydantic import Field
from schemas.common import BoundedCode64, BoundedText120, MonthStr, StrictApiModel

from auth import AuthClaims, require_auth
from composition import build_store_pnl_service
from permissions import can_access_management, require_privileged_access
from privileged_access import STORE_PNL_ACCESS_GROUPS_ENV, has_configured_group
from services.store_pnl import StorePnlService

router = APIRouter(prefix="/api/store-pnl", tags=["store-pnl"])


class PnlMetricsResponse(StrictApiModel):
    revenue: Decimal
    cogs: Decimal
    gross_margin: Decimal
    operating_costs: Decimal
    ebitda: Decimal
    depreciation: Decimal
    ebit: Decimal


class PnlMonthResponse(StrictApiModel):
    month: str
    has_actual: bool
    has_estimated: bool


class PnlPermissionsResponse(StrictApiModel):
    can_view: bool


class PnlMonthsResponse(StrictApiModel):
    months: list[PnlMonthResponse]


class PnlStoreOptionResponse(StrictApiModel):

    company_name: str
    site_code: str
    location: str
    regional: str | None = None
    scope_company: str | None = None


class PnlStoresResponse(StrictApiModel):
    stores: list[PnlStoreOptionResponse]


class PnlRegionsResponse(StrictApiModel):
    regions: list[str]


class PnlAnnualItemResponse(StrictApiModel):

    year: str
    store_count: int
    month_count: int
    is_estimated: bool
    revenue: Decimal
    cogs: Decimal
    gross_margin: Decimal
    operating_costs: Decimal
    ebitda: Decimal
    depreciation: Decimal
    ebit: Decimal


class PnlAnnualResponse(StrictApiModel):
    annual: list[PnlAnnualItemResponse]


class PnlMonthlyItemResponse(StrictApiModel):

    month: str
    is_estimated: bool
    revenue: Decimal
    cogs: Decimal
    gross_margin: Decimal
    operating_costs: Decimal
    ebitda: Decimal
    depreciation: Decimal
    ebit: Decimal


class PnlStoreResponse(StrictApiModel):

    company: str
    site_code: str
    source_site_code: str
    location: str
    regional: str | None = None
    has_estimates: bool
    revenue: Decimal
    cogs: Decimal
    gross_margin: Decimal
    operating_costs: Decimal
    ebitda: Decimal
    depreciation: Decimal
    ebit: Decimal


class PnlReconciliationResponse(StrictApiModel):
    month: str
    pnl_revenue: Decimal
    retail_sales_gross: Decimal
    retail_sales_net: Decimal
    difference_to_net: Decimal
    pnl_to_net_sales_pct: Decimal | None = None


class PnlOverviewResponse(StrictApiModel):

    start_month: str
    end_month: str
    company: str | None = None
    site_code: str | None = None
    site_company: str | None = None
    regional: str | None = None
    summary: PnlMetricsResponse
    monthly: list[PnlMonthlyItemResponse] = Field(default_factory=list)
    categories: dict[str, Decimal] = Field(default_factory=dict)
    stores: list[PnlStoreResponse] = Field(default_factory=list)
    reconciliation: list[PnlReconciliationResponse] = Field(default_factory=list)


def can_access_store_pnl(claims: AuthClaims) -> bool:
    return (
        can_access_management(claims)
        and has_configured_group(claims.groups, STORE_PNL_ACCESS_GROUPS_ENV)
    )


def require_store_pnl_owner(
    request: Request,
    claims: AuthClaims = Depends(require_auth),
) -> AuthClaims:
    return require_privileged_access(
        request=request,
        claims=claims,
        allowed=can_access_store_pnl(claims),
        resource="store_pnl",
        detail="Accesul la P&L nu este disponibil pentru acest utilizator.",
        fallback_route="/api/store-pnl",
    )


def parse_month(value: MonthStr) -> date:
    year, month = (int(part) for part in value.split("-"))
    return date(year, month, 1)


get_service = build_store_pnl_service


@router.get("/permissions", response_model=PnlPermissionsResponse)
async def pnl_permissions(claims: AuthClaims = Depends(require_auth)) -> dict[str, bool]:
    """Capability display endpoint; it intentionally does not emit audit events."""
    return {"can_view": can_access_store_pnl(claims)}


@router.get("/months", response_model=PnlMonthsResponse)
async def months(
    _claims: AuthClaims = Depends(require_store_pnl_owner),
    service: StorePnlService = Depends(get_service),
):
    return {"months": await service.months()}


def validate_company(company: str | None) -> None:
    if company not in (None, "Mobicell", "Mobiup"):
        raise HTTPException(status_code=422, detail="Companie P&L invalida.")


@router.get("/stores", response_model=PnlStoresResponse)
async def stores(
    company: BoundedText120 | None = Query(default=None),
    regional: BoundedText120 | None = Query(default=None),
    _claims: AuthClaims = Depends(require_store_pnl_owner),
    service: StorePnlService = Depends(get_service),
):
    validate_company(company)
    return {"stores": await service.stores(company, regional)}


@router.get("/regions", response_model=PnlRegionsResponse)
async def regions(
    company: BoundedText120 | None = Query(default=None),
    _claims: AuthClaims = Depends(require_store_pnl_owner),
    service: StorePnlService = Depends(get_service),
):
    validate_company(company)
    return {"regions": await service.regions(company)}


@router.get("/annual", response_model=PnlAnnualResponse)
async def annual(
    company: BoundedText120 | None = Query(default=None),
    site_code: BoundedCode64 | None = Query(default=None),
    site_company: BoundedText120 | None = Query(default=None),
    regional: BoundedText120 | None = Query(default=None),
    _claims: AuthClaims = Depends(require_store_pnl_owner),
    service: StorePnlService = Depends(get_service),
):
    validate_company(company)
    validate_company(site_company)
    return {"annual": await service.annual(company, site_code, site_company, regional)}


@router.get("/overview", response_model=PnlOverviewResponse)
async def overview(
    start_month: MonthStr,
    end_month: MonthStr,
    company: BoundedText120 | None = Query(default=None),
    site_code: BoundedCode64 | None = Query(default=None),
    site_company: BoundedText120 | None = Query(default=None),
    regional: BoundedText120 | None = Query(default=None),
    _claims: AuthClaims = Depends(require_store_pnl_owner),
    service: StorePnlService = Depends(get_service),
):
    validate_company(company)
    validate_company(site_company)
    start, end = parse_month(start_month), parse_month(end_month)
    if start > end:
        raise HTTPException(status_code=422, detail="Intervalul P&L este inversat.")
    return await service.overview(start, end, company, site_code, site_company, regional)
